File: platform/macos/Bridge/NativeValues.swift
// Copyright 2026 TidyVNC contributors. Licensed under GPL-2.0-or-later.
import Foundation
import CoreGraphics
import TidyVNC

public enum NativeStatus: UInt32, Sendable {
  case ok = 0, noChange = 1, pending = 2
  case invalidArgument = 10, abiMismatch, unsupported, invalidHandle, wrongHandleType
  case resourceLimit, outOfMemory, stale, notConnected, closing, busy, queueFull
  case cancelled, notPending, failed, internalFailure, viewOnly, unfocused, disabled, echo
}

public struct NativeError: Error, Sendable, Equatable, CustomStringConvertible {
  public let status: NativeStatus
  public let domain: UInt32
  public let detail: UInt32
  public let nativeCode: Int32
  public let description: String
  init(_ value: tidyvnc_error) {
    status = NativeStatus(rawValue: value.code) ?? .internalFailure
    domain = value.domain; detail = value.detail; nativeCode = value.native_error
    var text = value.message
    description = withUnsafeBytes(of: &text) { String(decoding: $0.prefix(while: { $0 != 0 }), as: UTF8.self) }
  }
  init(_ status: NativeStatus, _ message: String) {
    self.status = status; domain = UInt32(TIDYVNC_DOMAIN_BRIDGE)
    detail = 0; nativeCode = 0; description = message
  }
}

public enum NativeSessionState: UInt32, Sendable {
  case idle = 0, resolving, connecting, negotiating, authenticating, connected, disconnecting, closed, failed
}
public enum NativeEndReason: UInt32, Sendable {
  case none = 0, cancelled, peerClosed, promptTimeout, authenticationRejected, transport, protocolFailure
  case resource, internalFailure, eventOverflow, resolution, connection, resolutionTimeout, connectionTimeout
  case unsupportedEndpoint, invalidEndpoint
}
public struct NativeSnapshot: Sendable, Equatable {
  public let state: NativeSessionState
  public let endReason: NativeEndReason
  public let nativeCode: Int32
  public let width: UInt32, height: UInt32
  public let supportsResize: Bool, resizePending: Bool
  public let generation: UInt64, frames: UInt64, bells: UInt64
  public let information: NativeConnectionInformation?
  init(_ value: tidyvnc_snapshot, stateOverride: NativeSessionState? = nil, information: NativeConnectionInformation? = nil) {
    self.information = information
    state = stateOverride ?? NativeSessionState(rawValue: value.state) ?? .failed
    endReason = NativeEndReason(rawValue: value.end_reason) ?? .internalFailure
    nativeCode = value.native_error; width = value.width; height = value.height
    supportsResize = value.supports_resize != 0; resizePending = value.resize_pending != 0
    generation = value.generation; frames = value.frames; bells = value.bells
  }
}
extension tidyvnc_connection_info: ABIValue {}
public struct NativeConnectionInformation: Sendable, Equatable {
  public let generation: UInt64, frames: UInt64, bitsPerSecond: UInt64
  public let width: UInt32, height: UInt32, protocolMajor: UInt32, protocolMinor: UInt32, securityType: UInt32
  public let credentialsSecure: Bool, nameTruncated: Bool
  public let requestedEncoding: Int32, lastEncoding: Int32
  public let desktopName: String, pixelFormat: String, securityName: String
  public let requestedEncodingName: String, lastEncodingName: String
  init(_ value: tidyvnc_connection_info) {
    func text<T>(_ bytes: T) -> String { withUnsafeBytes(of: bytes) { String(decoding: $0.prefix(while: { $0 != 0 }), as: UTF8.self) } }
    generation = value.snapshot.generation; frames = value.snapshot.frames; bitsPerSecond = value.bits_per_second
    width = value.snapshot.width; height = value.snapshot.height; protocolMajor = value.protocol_major; protocolMinor = value.protocol_minor
    securityType = value.security_type; credentialsSecure = value.credentials_secure != 0; nameTruncated = value.name_truncated != 0
    requestedEncoding = value.requested_encoding; lastEncoding = value.last_encoding
    desktopName = text(value.desktop_name); pixelFormat = text(value.pixel_format); securityName = text(value.security_name)
    requestedEncodingName = text(value.requested_encoding_name); lastEncodingName = text(value.last_encoding_name)
  }
  // Deliberately exclude remote names, endpoints, paths and authentication data.
  public var redactedDiagnostics: String {
    """
    TidyVNC connection diagnostics
    Protocol: RFB \(protocolMajor).\(protocolMinor)
    Security: \(securityName) (\(securityType))
    Desktop size: \(width) × \(height)
    Pixel format: \(pixelFormat)
    Requested encoding: \(requestedEncodingName)
    Last used encoding: \(lastEncoding < 0 ? "Not received" : lastEncodingName)
    Line speed estimate: \(frames == 0 ? "Not sampled" : "\(bitsPerSecond / 1000) kbit/s")
    Frames received: \(frames)
    Endpoint and desktop name omitted.
    """
  }
}
public struct NativeOperation: Sendable, Equatable {
  public let id: UInt64, generation: UInt64
}
public struct NativeCompletion: Sendable, Equatable {
  public let operation: NativeOperation
  public let snapshot: NativeSnapshot
}
public struct NativeCommandFailure: Error, Sendable, Equatable {
  public enum Result: UInt32, Sendable { case succeeded = 0, cancelled, failed }
  public enum Reason: UInt32, Sendable { case none = 0, timedOut, serverRejected }
  public let operation: NativeOperation
  public let result: Result
  public let reason: Reason
  public let nativeResult: UInt32
  public let snapshot: NativeSnapshot
}

public struct NativeSessionConfiguration: Sendable {
  // nil uses the shared viewer default; explicit zero disables motion delay.
  public var pointerEventIntervalMilliseconds: UInt32? = nil
  public var pointerEventIntervalSource: NativeOptionSource? = nil
  // Incoming wire/decompressed clipboard bytes; independent of UTF-8 retention budgets.
  public var maxCutText: UInt32? = nil
  public var maxCutTextSource: NativeOptionSource? = nil
  public var windowStartupPolicy: NativeWindowStartupPolicy = .builtIn
  public var windowStartupSources: [NativeWindowStartupOption:NativeOptionSource] = [:]
  public var networkPolicy: NativeNetworkPolicy = .builtIn
  public var networkSources: [NativeNetworkOption:NativeOptionSource] = [:]
  public var fullscreenPolicy: NativeFullscreenPolicy = .builtIn
  public var fullscreenSources: [NativeFullscreenOption:NativeOptionSource] = [:]
  public var resizePolicy: NativeRemoteResizePolicy = .builtIn
  public var resizeSources: [NativeResizeOption:NativeOptionSource] = [:]
  public var shared = false, reconnectOnError = true
  public var sharedSource: NativeOptionSource = .compiled, reconnectSource: NativeOptionSource = .compiled
  // nil preserves a desktop host's explicit initial rendering configuration.
  public var scaling: NativeScaling? = nil
  public var scalingSources: [NativeScalingOption: NativeOptionSource] = [:]
  public var input = NativeInputSettings()
  public var inputSources: [NativeInputOption: NativeOptionSource] = [:]
  public var encoding: NativeEncodingOptions? = nil
  // nil snapshots the core's compiled defaults; [] explicitly denies all types.
  public var securityTypes: [UInt32]? = nil
  public var securitySource: NativeOptionSource? = nil
  public var promptTimeoutMilliseconds: UInt32 = 60_000
  public var eventCapacity: UInt32 = 128
  public var commandCapacity: UInt32 = 32
  public var framebufferBytes: UInt64 = 64 * 1024 * 1024
  public var publicationBytes: UInt64 = 128 * 1024 * 1024
  public var tlsPrioritySource: NativeOptionSource? = nil
  public var tlsPriority = "", caFile = "", crlFile = ""
  public var clipboardSend = true, clipboardReceive = true
  public init() {}
}

// Immutable ownership is safe across executors: the C registry and leases are
// thread-safe, and this class never mutates or publishes a borrowed pointer.
final class NativeHandle: @unchecked Sendable {
  let raw: UInt64
  init(adopting raw: UInt64) { precondition(raw != 0); self.raw = raw }
  convenience init(retaining raw: UInt64) throws {
    try checked { tidyvnc_retain(raw, $0) }; self.init(adopting: raw)
  }
  deinit { _ = tidyvnc_release(raw, nil) }
}

protocol ABIValue { init(); var size: UInt32 { get set }; var version: UInt32 { get set } }
extension tidyvnc_error: ABIValue {}
extension tidyvnc_runtime_options: ABIValue {}
extension tidyvnc_session_options: ABIValue {}
extension tidyvnc_connect_options: ABIValue {}
extension tidyvnc_operation: ABIValue {}
extension tidyvnc_snapshot: ABIValue {}
extension tidyvnc_event: ABIValue {}
extension tidyvnc_view_update: ABIValue {}
extension tidyvnc_image_info: ABIValue {}
extension tidyvnc_prompt_info: ABIValue {}
extension tidyvnc_callbacks: ABIValue {}
func abi<T: ABIValue>(_ type: T.Type) -> T {
  var value = T(); value.size = UInt32(MemoryLayout<T>.size); value.version = UInt32(TIDYVNC_ABI_VERSION); return value
}
@discardableResult
func checked(allowing: Set<NativeStatus> = [.ok], _ call: (UnsafeMutablePointer<tidyvnc_error>) -> UInt32) throws -> NativeStatus {
  var error = abi(tidyvnc_error.self)
  let code = call(&error)
  guard let status = NativeStatus(rawValue: code), allowing.contains(status) else { throw NativeError(error) }
  return status
}
func copyBytes(_ bytes: tidyvnc_bytes) throws -> Data {
  guard bytes.length <= UInt64(Int.max), bytes.data != nil || bytes.length == 0 else {
    throw NativeError(.internalFailure, "Invalid owned payload span")
  }
  guard bytes.length != 0 else { return Data() }
  return Data(bytes: bytes.data!, count: Int(bytes.length))
}
func withText<T>(_ text: String, _ body: (tidyvnc_bytes) throws -> T) rethrows -> T {
  try Array(text.utf8).withUnsafeBufferPointer { try body(tidyvnc_bytes(data: $0.baseAddress, length: UInt64($0.count))) }
}

public struct NativePixelRect: Sendable, Hashable {
  public let x: UInt32, y: UInt32, width: UInt32, height: UInt32
  public init(x: UInt32, y: UInt32, width: UInt32, height: UInt32) {
    self.x = x; self.y = y; self.width = width; self.height = height
  }
}

public final class NativeImage: @unchecked Sendable {
  public enum Format: UInt32, Sendable { case bgra8 = 1, rgba8 = 2 }
  public enum Alpha: UInt32, Sendable { case opaque = 1, straight = 2, premultiplied = 3 }
  let handle: NativeHandle
  public let width: UInt32, height: UInt32, stride: UInt64
  public let generation: UInt64, sizeGeneration: UInt64, sequence: UInt64
  public let previousSequence: UInt64
  public let damage: NativePixelRect
  let streamID: UUID
  public let hotspotX: UInt32, hotspotY: UInt32
  public let format: Format, alpha: Alpha
  init(owning owner: NativeHandle, previousSequence: UInt64 = 0, damage: NativePixelRect? = nil, streamID: UUID = UUID()) throws {
    var info = abi(tidyvnc_image_info.self)
    try checked { tidyvnc_image_get(owner.raw, &info, $0) }
    guard let format = Format(rawValue: info.format), let alpha = Alpha(rawValue: info.alpha),
          info.origin == UInt32(TIDYVNC_ORIGIN_TOP_LEFT) else { throw NativeError(.unsupported, "Unsupported image format") }
    self.previousSequence = previousSequence; self.streamID = streamID
    self.damage = damage ?? NativePixelRect(x: 0, y: 0, width: info.width, height: info.height)
    handle = owner; width = info.width; height = info.height; stride = info.stride
    generation = info.generation; sizeGeneration = info.size_generation; sequence = info.sequence
    hotspotX = info.hotspot_x; hotspotY = info.hotspot_y; self.format = format; self.alpha = alpha
  }
  // The retained image itself is zero-copy. A client needing Swift-owned bytes
  // explicitly requests a copy; no unsafe borrowed pixel span enters UI state.
  public func copyPixels() throws -> Data {
    try withExtendedLifetime(handle) {
      var info = abi(tidyvnc_image_info.self)
      try checked { tidyvnc_image_get(handle.raw, &info, $0) }
      return try copyBytes(info.pixels)
    }
  }
  // The provider retains the C lease until Core Graphics releases the image,
  // including after view removal, resize or session/runtime shutdown.
  public func makeCGImage() throws -> CGImage {
    var info = abi(tidyvnc_image_info.self)
    try checked { tidyvnc_image_get(handle.raw, &info, $0) }
    guard let pixels = info.pixels.data, info.pixels.length <= UInt64(Int.max), stride <= UInt64(Int.max) else {
      throw NativeError(.internalFailure, "Invalid image storage")
    }
    let retained = Unmanaged.passRetained(handle)
    guard let provider = CGDataProvider(dataInfo: retained.toOpaque(), data: pixels, size: Int(info.pixels.length), releaseData: { context, _, _ in
      if let context { Unmanaged<NativeHandle>.fromOpaque(context).release() }
    }) else { retained.release(); throw NativeError(.outOfMemory, "Could not retain image data") }
    let alphaInfo: CGImageAlphaInfo
    if format == .bgra8 {
      alphaInfo = alpha == .opaque ? .noneSkipFirst : alpha == .straight ? .first : .premultipliedFirst
    } else {
      alphaInfo = alpha == .opaque ? .noneSkipLast : alpha == .straight ? .last : .premultipliedLast
    }
    let order: CGBitmapInfo = format == .bgra8 ? .byteOrder32Little : .byteOrder32Big
    guard let image = CGImage(width: Int(width), height: Int(height), bitsPerComponent: 8, bitsPerPixel: 32,
      bytesPerRow: Int(stride), space: CGColorSpace(name: CGColorSpace.sRGB)!,
      bitmapInfo: order.union(CGBitmapInfo(rawValue: alphaInfo.rawValue)), provider: provider,
      decode: nil, shouldInterpolate: false, intent: .defaultIntent) else { throw NativeError(.failed, "Could not create desktop image") }
    return image
  }
}

public struct NativePrompt: Sendable, Equatable, Identifiable {
  public enum Kind: UInt32, Sendable { case credentials = 1, certificate, hostKey }
  public let id: UInt64, generation: UInt64
  public let kind: Kind
  public let secure: Bool, usernameRequired: Bool
  // The core's isSecure policy assesses credential protection. It is not a
  // transport-encryption flag (anonymous TLS is false; security modes differ).
  public var credentialProtectionMessage: String {
    secure ? "The negotiated authentication method protects your credentials." :
      "The negotiated authentication method may not adequately protect your credentials."
  }
  public let securityType: UInt32
  public let certificateStatus: UInt32
  public let serverName: String, fingerprint: String
  public let identity: Data
  init(adopting raw: UInt64) throws {
    let owner = NativeHandle(adopting: raw)
    self = try withExtendedLifetime(owner) {
      var info = abi(tidyvnc_prompt_info.self)
      try checked { tidyvnc_prompt_get(owner.raw, &info, $0) }
      var securityType: UInt32 = 0
      try checked { tidyvnc_prompt_security_type(owner.raw, &securityType, $0) }
      guard let kind = Kind(rawValue: info.kind) else { throw NativeError(.unsupported, "Unsupported authentication prompt") }
      return NativePrompt(id: info.id, generation: info.generation, kind: kind, secure: info.secure != 0, securityType: securityType,
        usernameRequired: info.username_required != 0, certificateStatus: info.certificate_status,
        serverName: String(decoding: try copyBytes(info.server_name), as: UTF8.self),
        fingerprint: String(decoding: try copyBytes(info.fingerprint), as: UTF8.self), identity: try copyBytes(info.identity))
    }
  }
  init(id: UInt64, generation: UInt64, kind: Kind, secure: Bool, securityType: UInt32 = 0, usernameRequired: Bool,
               certificateStatus: UInt32, serverName: String, fingerprint: String, identity: Data) {
    self.id = id; self.generation = generation; self.kind = kind; self.secure = secure
    self.securityType = securityType
    self.usernameRequired = usernameRequired; self.certificateStatus = certificateStatus
    self.serverName = serverName; self.fingerprint = fingerprint; self.identity = identity
  }
}

File: platform/windows/TidyVNC.Native/Bridge/NativeValues.cs
// Copyright 2026 TidyVNC contributors. Licensed under GPL-2.0-or-later.
using System.Security.Cryptography;
using TidyVNC.Native.Interop;

namespace TidyVNC.Native;

public enum NativeSessionState : uint
{
    Idle = 0, Resolving, Connecting, Negotiating, Authenticating, Connected, Disconnecting, Closed, Failed,
}

public enum NativeEndReason : uint
{
    None = 0, Cancelled, PeerClosed, PromptTimeout, AuthenticationRejected, Transport, ProtocolFailure,
    Resource, InternalFailure, EventOverflow, Resolution, Connection, ResolutionTimeout, ConnectionTimeout,
    UnsupportedEndpoint, InvalidEndpoint,
}

/// <summary>One published session state (NativeValues.swift NativeSnapshot).</summary>
public sealed record NativeSnapshot(
    NativeSessionState State, NativeEndReason EndReason, int NativeCode, uint Width, uint Height,
    bool SupportsResize, bool ResizePending, ulong Generation, ulong Frames, ulong Bells,
    NativeConnectionInformation? Information = null)
{
    internal static NativeSnapshot From(in tidyvnc_snapshot value, NativeSessionState? stateOverride = null,
                                        NativeConnectionInformation? information = null)
        => new(stateOverride ?? (Enum.IsDefined((NativeSessionState)value.state) ? (NativeSessionState)value.state : NativeSessionState.Failed),
               Enum.IsDefined((NativeEndReason)value.end_reason) ? (NativeEndReason)value.end_reason : NativeEndReason.InternalFailure,
               value.native_error, value.width, value.height, value.supports_resize != 0, value.resize_pending != 0,
               value.generation, value.frames, value.bells, information);
}

/// <summary>Negotiated connection details (NativeConnectionInformation in Swift).</summary>
public sealed record NativeConnectionInformation(
    ulong Generation, ulong Frames, ulong BitsPerSecond, uint Width, uint Height, uint ProtocolMajor, uint ProtocolMinor,
    uint SecurityType, bool CredentialsSecure, bool NameTruncated, int RequestedEncoding, int LastEncoding,
    string DesktopName, string PixelFormat, string SecurityName, string RequestedEncodingName, string LastEncodingName)
{
    internal static unsafe NativeConnectionInformation From(tidyvnc_connection_info value)
        => new(value.snapshot.generation, value.snapshot.frames, value.bits_per_second, value.snapshot.width, value.snapshot.height,
               value.protocol_major, value.protocol_minor, value.security_type, value.credentials_secure != 0, value.name_truncated != 0,
               value.requested_encoding, value.last_encoding,
               NativeText.Fixed(value.desktop_name, 1025), NativeText.Fixed(value.pixel_format, 128),
               NativeText.Fixed(value.security_name, 64), NativeText.Fixed(value.requested_encoding_name, 32),
               NativeText.Fixed(value.last_encoding_name, 32));

    /// <summary>Diagnostics safe to copy: no remote names, endpoints, paths or authentication data.</summary>
    public string RedactedDiagnostics =>
        $"""
        TidyVNC connection diagnostics
        Protocol: RFB {ProtocolMajor}.{ProtocolMinor}
        Security: {SecurityName} ({SecurityType})
        Desktop size: {Width} × {Height}
        Pixel format: {PixelFormat}
        Requested encoding: {RequestedEncodingName}
        Last used encoding: {(LastEncoding < 0 ? "Not received" : LastEncodingName)}
        Line speed estimate: {(Frames == 0 ? "Not sampled" : $"{BitsPerSecond / 1000} kbit/s")}
        Frames received: {Frames}
        Endpoint and desktop name omitted.
        """;
}

public readonly record struct NativeOperation(ulong Id, ulong Generation);

public sealed record NativeCompletion(NativeOperation Operation, NativeSnapshot Snapshot);

/// <summary>A command the core admitted but did not complete successfully.</summary>
public sealed class NativeCommandFailure : Exception
{
    public enum ResultKind : uint { Succeeded = 0, Cancelled, Failed }
    public enum FailureReason : uint { None = 0, TimedOut, ServerRejected }

    public NativeOperation Operation { get; }
    public ResultKind Result { get; }
    public FailureReason Reason { get; }
    public uint NativeResult { get; }
    public NativeSnapshot Snapshot { get; }

    internal NativeCommandFailure(NativeOperation operation, ResultKind result, FailureReason reason, uint nativeResult, NativeSnapshot snapshot)
        : base($"Operation {result} ({reason})")
    {
        Operation = operation; Result = result; Reason = reason; NativeResult = nativeResult; Snapshot = snapshot;
    }
}

public readonly record struct NativePixelRect(uint X, uint Y, uint Width, uint Height);

/// <summary>
/// A retained, immutable frame or cursor lease (NativeImage in Swift). Pixels
/// stay in the core; renderers borrow them through the lease while it lives.
/// </summary>
public sealed class NativeImage : IDisposable
{
    public enum PixelFormat : uint { Bgra8 = 1, Rgba8 = 2 }
    public enum AlphaMode : uint { Opaque = 1, Straight = 2, Premultiplied = 3 }

    internal NativeHandle Handle { get; }
    public uint Width { get; }
    public uint Height { get; }
    public ulong Stride { get; }
    public ulong Generation { get; }
    public ulong SizeGeneration { get; }
    public ulong Sequence { get; }
    public ulong PreviousSequence { get; }
    public NativePixelRect Damage { get; }
    public Guid StreamId { get; }
    public uint HotspotX { get; }
    public uint HotspotY { get; }
    public PixelFormat Format { get; }
    public AlphaMode Alpha { get; }

    internal unsafe NativeImage(NativeHandle owner, ulong previousSequence = 0, NativePixelRect? damage = null, Guid? streamId = null)
    {
        var info = Abi.Init<tidyvnc_image_info>();
        var error = Abi.Init<tidyvnc_error>();
        Abi.Check(NativeMethods.tidyvnc_image_get(owner.Raw, &info, &error), &error);
        if (!Enum.IsDefined((PixelFormat)info.format) || !Enum.IsDefined((AlphaMode)info.alpha) ||
            info.origin != Tidyvnc.TIDYVNC_ORIGIN_TOP_LEFT)
            throw new NativeError(NativeStatus.Unsupported, "Unsupported image format");
        Handle = owner; Width = info.width; Height = info.height; Stride = info.stride;
        Generation = info.generation; SizeGeneration = info.size_generation; Sequence = info.sequence;
        HotspotX = info.hotspot_x; HotspotY = info.hotspot_y;
        Format = (PixelFormat)info.format; Alpha = (AlphaMode)info.alpha;
        PreviousSequence = previousSequence; StreamId = streamId ?? Guid.NewGuid();
        Damage = damage ?? new NativePixelRect(0, 0, info.width, info.height);
    }

    /// <summary>Borrowed pixels, valid while this lease is alive. Callers keep the lease referenced.</summary>
    public unsafe ReadOnlySpan<byte> BorrowPixels()
    {
        var info = Abi.Init<tidyvnc_image_info>();
        var error = Abi.Init<tidyvnc_error>();
        Abi.Check(NativeMethods.tidyvnc_image_get(Handle.Raw, &info, &error), &error);
        if (info.pixels.data == null || info.pixels.length > int.MaxValue)
            throw new NativeError(NativeStatus.InternalFailure, "Invalid image storage");
        return new ReadOnlySpan<byte>(info.pixels.data, (int)info.pixels.length);
    }

    public byte[] CopyPixels() => BorrowPixels().ToArray();

    private NativeImage(NativeImage source, NativeHandle handle)
    {
        Handle = handle; Width = source.Width; Height = source.Height; Stride = source.Stride;
        Generation = source.Generation; SizeGeneration = source.SizeGeneration; Sequence = source.Sequence;
        PreviousSequence = source.PreviousSequence; Damage = source.Damage; StreamId = source.StreamId;
        HotspotX = source.HotspotX; HotspotY = source.HotspotY; Format = source.Format; Alpha = source.Alpha;
    }

    /// <summary>
    /// An independent reference to the same lease. Each holder (session state,
    /// a render worker) disposes its own, so pixels are freed deterministically.
    /// </summary>
    public NativeImage Clone() => new(this, NativeHandle.Retain(Handle.Raw));

    public void Dispose() => Handle.Dispose();
}

/// <summary>An authentication or trust request (NativePrompt in Swift).</summary>
public sealed record NativePrompt(
    ulong Id, ulong Generation, NativePrompt.PromptKind Kind, bool Secure, uint SecurityType, bool UsernameRequired,
    uint CertificateStatus, string ServerName, string Fingerprint, byte[] Identity)
{
    public enum PromptKind : uint { Credentials = 1, Certificate = 2, HostKey = 3 }

    internal static unsafe NativePrompt Adopt(ulong raw)
    {
        using var owner = NativeHandle.Adopt(raw);
        var info = Abi.Init<tidyvnc_prompt_info>();
        var error = Abi.Init<tidyvnc_error>();
        Abi.Check(NativeMethods.tidyvnc_prompt_get(owner.Raw, &info, &error), &error);
        uint securityType = 0;
        Abi.Check(NativeMethods.tidyvnc_prompt_security_type(owner.Raw, &securityType, &error), &error);
        if (!Enum.IsDefined((PromptKind)info.kind)) throw new NativeError(NativeStatus.Unsupported, "Unsupported authentication prompt");
        return new NativePrompt(info.id, info.generation, (PromptKind)info.kind, info.secure != 0, securityType,
            info.username_required != 0, info.certificate_status, NativeText.Decode(info.server_name),
            NativeText.Decode(info.fingerprint), NativeText.Copy(info.identity));
    }

    /// <summary>SHA-256 of the raw identity, upper-case hex pairs separated by colons.</summary>
    public string? Sha256Fingerprint => Identity.Length is > 0 and <= 65536
        ? string.Join(':', SHA256.HashData(Identity).Select(b => b.ToString("X2", System.Globalization.CultureInfo.InvariantCulture)))
        : null;

    public override string ToString() => $"NativePrompt({Kind}, <redacted>)";
}
